lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -75,49 +75,103 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
+  /// substitutions.
+  DenseMap<Attribute, unsigned> symbolMap;
+
+  /// Symbol substitutions for the JSON verbatim op.
+  SmallVector<Attribute> symbols;
+
+  /// Get a symbol index and update symbol datastructures.
+  unsigned getSymbolIndex(Attribute attr) {
+    auto iterator = symbolMap.find(attr);
+    if (iterator != symbolMap.end())
+      return iterator->getSecond();
+
+    auto idx = symbolIdx++;
+    symbolMap.insert({attr, idx});
+    symbols.push_back(attr);
+
+    return idx;
+  }
+
+  /// Convert an annotation path to a string with symbol substitutions.
+  void buildTarget(AnnoPathValue &path, SmallString<64> &newTarget) {
+
+    auto addSymbol = [&](Attribute attr) -> void {
+      newTarget.append("{{");
+      newTarget.append(std::to_string(getSymbolIndex(attr)));
+      newTarget.append("}}");
+    };
+
+    newTarget.append("~");
+    newTarget.append(
+        path.ref.getModule()->getParentOfType<CircuitOp>().getName());
     newTarget.append("|");
 
-    if (auto nla = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
-      hw::HierPathOp path = nlaTable->getNLA(nla.getAttr());
-      for (auto part : path.getNamepath().getValue().drop_back()) {
-        auto inst = cast<hw::InnerRefAttr>(part);
-        newTarget.append(inst.getModule());
-        newTarget.append("/");
-        newTarget.append(inst.getName());
-        newTarget.append(":");
-      }
+    if (path.isLocal()) {
+      addSymbol(
+          FlatSymbolRefAttr::get(path.ref.getModule().getModuleNameAttr()));
+    } else {
+      addSymbol(FlatSymbolRefAttr::get(path.instances.front()
+                                           ->getParentOfType<FModuleLike>()
+                                           .getModuleNameAttr()));
+    }
+
+    for (auto inst : path.instances) {
+      newTarget.append("/");
+      addSymbol(hw::InnerRefAttr::get(
+          inst->getParentOfType<FModuleLike>().getModuleNameAttr(),
+          inst.getInnerSymAttr().getSymName()));
+      newTarget.append(":");
+      addSymbol(inst.getModuleNameAttr());
     }
-    newTarget.append(module.getModuleName());
-    newTarget.append(">");
 
-    newTarget.append(name);
+    // If this targets a module or an instance, then we're done.  There is no
+    // "reference" part of the FIRRTL target.
+    if (path.ref.isa<OpAnnoTarget>() &&
+        path.isOpOfType<FModuleOp, FExtModuleOp, InstanceOp>())
+      return;
 
-    // Add information if this is an aggregate.
-    auto targetFieldID = anno.getFieldID();
+    newTarget.append(">");
+    auto innerSym =
+        TypeSwitch<AnnoTarget, hw::InnerSymAttr>(path.ref)
+            .Case<PortAnnoTarget>([&](PortAnnoTarget portTarget) {
+              return portTarget.getModule().getPortSymbolAttr(
+                  portTarget.getPortNo());
+            })
+            .Case<OpAnnoTarget>([&](OpAnnoTarget opTarget) {
+              return opTarget.getOp()->getAttrOfType<hw::InnerSymAttr>(
```

Comment:
Please don't introduce new raw uses of "inner_sym" by name.  Will be making pass fixing this in HW "soon", looks like here and a few other places we use raw name (most were removed previously).

Reply:
Fair. For some reason `getOrAddInnerSym` didn't occur to me even though it's used in the pass already. 😕  Will fix.